plugins/samples/content_injection/plugin.rs

```diff
@@ -0,0 +1,180 @@
+use lol_html::*;
+use proxy_wasm::traits::*;
+use proxy_wasm::types::*;
+use std::cell::RefCell;
+use std::error::Error;
+use std::rc::Rc;
+
+proxy_wasm::main! {{
+    proxy_wasm::set_log_level(LogLevel::Trace);
+    proxy_wasm::set_http_context(|_, _| -> Box<dyn HttpContext> { MyHttpContext::new()});
+}}
+
+struct MyOutputSink {
+    // Stores HTML as the rewriter parses and modifies HTML.
+    // Only stores sections of HTML that have not yet been seen back to client.
+    // After sending data to client, we clear output to avoid unnecessary memory
+    // growth.
+    // See comment in MyHttpContext about use of Rc<RefCell<T>>.
+    output_sink: Rc<RefCell<Vec<u8>>>,
+}
+
+impl OutputSink for MyOutputSink {
+    fn handle_chunk(&mut self, chunk: &[u8]) {
+        self.output_sink.borrow_mut().extend_from_slice(chunk);
+    }
+}
+
+struct MyHttpContext<'a> {
+    // Some member variables need to be wrapped in Rc<Refcell<T>> so that we can
+    // modify the variable in multiple places as well as share ownership.
+    // By using Rc, we are allowing multiple references to the same obeject to
+    // exist at the same time.
+    // By using Refcell, we are evaluating borrow checks at runtime, thus we are
+    // commiting to not modfying the variable in multiple places simultaneously,
+    // otherwise the code will panic and plugin will crash.
+
+    // Stores HTML as the rewriter parses and modifies HTML.
+    // Only stores sections of HTML that have not yet been seen back to client.
+    // After sending data to client, we clear output to avoid unnecessary memory
+    // growth.
+    output: Rc<RefCell<Vec<u8>>>,
+    // HtmlRewriter. Member of MyHttpContext a.k.a "StreamContext" so that the
+    // rewriter persists across multiple body callbacks.The rewriter parses
+    // and modifies chunks of Html. Behavior for Html modifications are defined
+    // via content_handlers. The rewriter only writes completed sections of Html to
+    // the output sink(e.g., when the rewriter parses "<di", nothing will be
+    // sent to the output sink. If the next chunk was "v> <h1>foo</h", the
+    // rewriter would then write "<div><h1>foo" to the output sink) unless
+    // HtmlRewriter::end() is called. HtmlRewriter::end() results in all
+    // uncompleted Html being buffered in rewriter to be sent to output sink as
+    // if it were plain text.
+    rewriter: Option<HtmlRewriter<'a, MyOutputSink>>,
+    // True when plugin has added script to <head>.
+    completed: Rc<RefCell<bool>>,
+}
+
+impl<'a> MyHttpContext<'a> {
+    pub fn new() -> Box<MyHttpContext<'a>> {
+        let output = Rc::new(RefCell::new(Vec::new()));
+        let completed = Rc::new(RefCell::new(false));
+        Box::new(MyHttpContext {
+            output: output.clone(),
+            completed: completed.clone(),
+            rewriter: Some(HtmlRewriter::new(
+                Settings {
+                    element_content_handlers: vec![element!("head", move |el| {
```

Comment:
Can you add a comment or extract this into a helper function that makes it clear that this is where a customer would add logic for what to inject?

It seems like it should be in the on_response_body handler, but it's actually in the constructor due to the design of the HTML Rewriter.

Reply:
Done